### DISCUSSION TITLE
🗺 Roadmap: Features & Docs

### DISCUSSION THREAD

**Discussion by davidkpiano** [statelyai/xstate#742] [General]
## Proposed Features & Enhancements

- [ ] Relative deep targets https://github.com/davidkpiano/xstate/issues/554
- [x] Generic XState Dev Tool support https://github.com/davidkpiano/xstate/issues/709
- [x] (V5) Combine activities & invoked services https://github.com/davidkpiano/xstate/issues/628
- [ ] Expand `@xstate/test` plan strategies https://github.com/davidkpiano/xstate/issues/694
- [x] Microstep listener https://github.com/davidkpiano/xstate/issues/692 https://github.com/davidkpiano/xstate/issues/251
- [ ] Relative paths for in-state guards https://github.com/davidkpiano/xstate/issues/349
- [ ] `@xstate/react` snapshot testing https://github.com/davidkpiano/xstate/issues/432
- [x] Deep `initial` targets https://github.com/davidkpiano/xstate/issues/675
- [x] Actions for `initial` (full config) https://github.com/davidkpiano/xstate/issues/466
- [ ] Abort controller support for `invoke` https://github.com/davidkpiano/xstate/issues/418
- [ ] `@xstate/scxml` SCXML import https://github.com/davidkpiano/xstate/issues/411 https://github.com/davidkpiano/xstate/issues/101
- [x] Split `StateMachine` and `StateNode` https://github.com/davidkpiano/xstate/issues/655
- [ ] Visualizer: cache-bust fetching of gists https://github.com/davidkpiano/xstate/issues/635
- [x] Expose invoked services https://github.com/davidkpiano/xstate/issues/424
- [ ] sketch.systems to XState https://github.com/davidkpiano/xstate/issues/479
- [x] Visualizer: support displaying multiple machines https://github.com/davidkpiano/xstate/issues/625
- [x] Visualizer: TypeScript support https://github.com/davidkpiano/xstate/issues/429
- [ ] XState Storybook addon https://github.com/davidkpiano/xstate/issues/30
- [ ] Design by contract https://github.com/davidkpiano/xstate/issues/183
- [ ] `@xstate/graph` Initial transition https://github.com/davidkpiano/xstate/issues/266
- [ ] Tooling for gradual statechart specification https://github.com/davidkpiano/xstate/issues/287
- [x] `state.done` https://github.com/davidkpiano/xstate/issues/328
- [ ] Visualizer: XState JSON support https://github.com/davidkpiano/xstate/issues/401
- [ ] Visualizer: SVG, ASCII output support https://github.com/davidkpiano/xstate/issues/90
- [ ] Context replace https://github.com/davidkpiano/xstate/issues/262
- [ ] Visualizer: validation https://github.com/davidkpiano/xstate/issues/122
- [ ] Visualizer: meta data https://github.com/davidkpiano/xstate/issues/447
- [ ] Restoring services https://github.com/davidkpiano/xstate/issues/332
- [ ] Visualizer: CLI tool https://github.com/davidkpiano/xstate/issues/528
- [ ] Array conditions https://github.com/davidkpiano/xstate/issues/414
- [ ] Visualize variables of custom guards #881 
- [ ] Visualizer: SSO without GitHub https://github.com/davidkpiano/xstate/issues/751
- [x] `@xstate/react` React-specific interpreter for concurrent mode https://github.com/davidkpiano/xstate/issues/1094
- [ ] Action arrays https://github.com/davidkpiano/xstate/issues/1104

## Planned Documentation
- [ ] Using `SimulatedClock` with `@xstate/test` https://github.com/davidkpiano/xstate/issues/642
- [ ] Transition execution sequence timing https://github.com/davidkpiano/xstate/issues/621
- [ ] Improve docs on invoking services https://github.com/davidkpiano/xstate/issues/268
- [ ] Pattern for next/back navigation https://github.com/davidkpiano/xstate/issues/188 https://github.com/davidkpiano/xstate/issues/481
- [ ] Using XState in other frameworks (Svelte, Angular, etc.) https://github.com/davidkpiano/xstate/issues/1149
- [ ] Stopping spawned actors https://github.com/davidkpiano/xstate/issues/1116
- [ ] When to use actors https://github.com/davidkpiano/xstate/issues/906

**Comment by CodingDive** [comment]
I'm most excited about the browser extension and visualizing actors 🎉

**Comment by CodingDive** [comment]
Are there any plans for a typed matching API? Whenever I'm refactoring my machines and move state nodes around, I usually need to Ctrl+Shift+F every moved state node to ensure I'm still matching against the correct state structure. It would be great if TypeScript could catch those sort of errors since I'm already defining a state schema for every machine. 😊 

Inspired by @christianalfoni video on how this works in overmind https://www.youtube.com/watch?v=uDS159mWJ58

**Comment by semopz** [comment]
Going forward, could you link us to to a relevant announcement as you complete these tasks? If the link is to a release notes section where multiple features have been introduced, it should be made clear which roadmap item the feature relates to. 

So when I visit this page, I'd like to know
1. Has the feature been implemented (available now via checkboxes 👍)
2. Deps for the feature (e.g., you need xstate v4.7+ AND @xstate/react v1.0+)
3. How to use the feature (examples, API docs link, etc...)
4. The name of the implemented feature. For example `Visualizer: support displaying multiple machines #625` will probably become the `Support for the Actors model in viz` feature (or something to that effect)

**Comment by ThaddeusJiang** [comment]
Hi @davidkpiano 
Could we writing a real world Testing Example.
like your talking https://www.youtube.com/watch?v=tpNmPKjPSFQ

**Comment by andria-dev** [comment]
Any news on "Abort controller support for invoke" being worked on as #418 has been closed, @davidkpiano?

**Reply by CodingDive** [reply]
Can't speak for David but I believe the tasks of this issue/discussion thread is the sole and primary way to track feature requests until they'll be implemented.

I also think AbortController support is expected to be supported out of the box in v5. 😍 I'd too be interested how the final API will look like. The best thing would be to just transition into another state node and have XState do the promise cleanup of a service `src` behind the scenes. However, since promises can't natively be cancelled, some additional code may be needed to cancel a particularly fetch request in user land.

**Reply by Andarist** [reply]
IMHO AbortController case will just require a special invoke/spawn wrapper - currently in v5 we have builtin `invokePromise`, `invokeCallback` etc. So we could quite easily introduce `invokeAbortablePromise` or something like that.

**Comment by danielkcz** [comment]
Is there some ETA on V5? It's just that we are starting a new project roughly in 2 weeks and it would be nice to go with that right away.

All are those roadmap items expected to be done before V5 hits or is there some plan for alpha version soon?

**Reply by davidkpiano** [reply]
> is there some plan for alpha version soon?

Yes, there is.

I think we can ship an alpha within the next two weeks (cc. @Andarist)

**Reply by danielkcz** [reply]
I don't want to be annoying or anything, but any chance to get alpha version out soon?

**Reply by Andarist** [reply]
We’ll talk this through with @davidkpiano and probably gonna release something - bear the mind its alpha status though 😉

**Comment by noor-tg** [comment]
hey, any ETA for the v5 ?

**Reply by davidkpiano** [reply]
It's still a slow-but-steady work in progress. Hope to have a release candidate out by December, though.

**Reply by danielkcz** [reply]
So not even some alpha unstable as I was asking 3 months ago here? :) Just to get the feel for changes, not some serious work.

**Reply by davidkpiano** [reply]
@FredyC Feel free to play with the `next` branch! It's fully working and available. It's just not on npm yet, but that shouldn't stop you from getting a feel for the changes.

Also, keep in mind that the definition of "3 months ago" has wildly changed in 2020. Might as well be 3 days ago.

**Comment by mikestopcontinues** [comment]
Are you doing anything to tackle the bundle size? For all the features, it's understandably heavy. But what if the baseline was similar to `fsm` (or less!), and you could opt into all of the additional features as plugins? If the plugins were limited to only those provided by xstate, you could still easily control ordering and adherence to the state charts spec. 

I worry that as the library grows, it's going to become so large that only really big apps (games, graphics editors, etc) can justify it.

**Reply by davidkpiano** [reply]
Yes, this is planned for v5, which will take a more modular approach.

**Reply by bmadone** [reply]
Hi @davidkpiano is 5 version still in development, or it abandoned?

**Reply by Andarist** [reply]
It is in development

**Reply by UnknownH** [reply]
Will V5 be fundamentally different to V4 in terms of operation and developer experience? And if that will be the case, is there any way we can start learning V5 now?
We'll be starting a new project soon and we are considering XState, but I wouldn't like to start learning and implementing it now, just to have it superseded in 4 or 5 months time by a fundamentally different library.

**Reply by Andarist** [reply]
> Will V5 be fundamentally different to V4 in terms of operation and developer experience?

Fundamental concepts will stay the same but, of course, we plan to improve the developer experience in many ways. So some API changes are expected - the core functionality etc should stay mostly the same though.

A good chunk of the currently implemented changes is mentioned [here](https://github.com/statelyai/xstate/blob/4dd06779c85fd06e740a524ee43eec4606240891/packages/core/CHANGELOG.md#500-alpha0) and [here](https://github.com/statelyai/xstate/blob/4dd06779c85fd06e740a524ee43eec4606240891/packages/core/CHANGELOG.md#500-alpha1)

**Reply by UnknownH** [reply]
Thank you for your reply. I presume that there is still no ETA for V5 right?

**Reply by Andarist** [reply]
The ETA is "when it's ready" 😉

**Reply by UnknownH** [reply]
Thank you @Andarist for the honest answer, I appreciate it. Unfortunately it will be very difficult for me to explain that level of uncertainty to the stakeholders, since big banks are extremely conservative. So as much as we'd like to use XState (since that would set precedent for other projects across the organisation), for the time being we will stick to well known and stable formulas. But for future projects, maybe when there is more visibility about the project, we will come back and check. See you then 🙂.

**Reply by mikestopcontinues** [reply]
I'm not sure that's the right way to look at it. You've got a stable, functional, exceptionally useful library available today. No need to wait for the next version to start, right? Enterprise likes nothing better than a track record, and xstate has it.

**Reply by UnknownH** [reply]
From previous experience there is nothing that large corporations hate more than 1) Framework migrations (ex from angular to react, or from class components to hooks). 2) Library adoption when another major version is to be released "soon". Reason being because an effort (=time =money) will need to be put into upgrading instead of focusing on features. Beyond that, and also from experience, when a framework/library operates in a "will be released when it's ready"-basis, that usually raises major red flags among stakeholders and upper management. In my opinion is not a bad way to operate, but maybe more for a startup environment 🙂.

**Reply by davidkpiano** [reply]
@UnknownH Thanks for the feedback. A more accurate timeline is this:

- XState v5 is currently in **alpha** (you can install it via `xstate@alpha`), but there are many major changes still expected.
- We hope to have an XState v5 beta released by the end of the year/early next year. There are a few big moving parts that need to go in for this; namely, invoking/spawning actor refactor, system architecture, input/output, type safety.
- XState v5 will not deviate much from XState v4, and there should be a very straightforward migration path.
- XState v4 will continue to be supported for a long time, especially since it has been around for 4+ years now and is used in production at many companies.
- XState v4 is currently in feature freeze. We will fix bugs and make improvements but the majority of development effort will be for v5.

**Reply by bmadone** [reply]
@davidkpiano it would be great, thanks!

**Reply by UnknownH** [reply]
Thank you very much @davidkpiano for bringing some clarity regarding the timelines! Should I understand then that, if we were to start implementing the project using XState 4, a very simple migration path (in terms of time) to XState 5 would be expected? Let's say one week or less, depending on the size of the project of course, but granted that we will be starting in the following weeks, and with Christmas right around the corner, there won't be much time to get really far... That would make it so much easy to explain and get approval from the upper echelon

**Reply by davidkpiano** [reply]
@UnknownH Yes, the migration path is expected to be straightforward, and we have talked about automating it with a codemod. The machine structure will be largely unchanged.

**Comment by NoriSte** [comment]
A generic question about this roadmap: it is still valid? 😊
I see the last time it's been updated is October 2020 (even if there are more recent comments in this discussion and even if the issues could have been updated more recently) 😊

**Reply by davidkpiano** [reply]
This roadmap isn't up to date - the most up to date future feature list/roadmap would be at https://feedback.stately.ai

**Reply by NoriSte** [reply]
Nice, thank you 😊
